### ISSUE TITLE
HMR with TypeScript causes infinite reloading loops

### ISSUE THREAD

**Issue by orta** [zeit/next.js#4285]
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues?q=is%3Aissue) of this repository and believe that this is not a duplicate.

👋 Hey everyone, I've been using next with TypeScript (was 6.0 canary, now with 6.0.1 canary) and HMR has been causing quite a lot of issues. I created an example repo to showcase it: nextjs-typescript-hrm-bug-example

Basically the default setup for HRM in nextjs means that making a change in a `*.tsx` file would cause an infinite loop with these errors

<img width="1091" alt="screen shot 2018-05-06 at 4 08 44 pm" src="https://user-images.githubusercontent.com/49038/39677627-59a67990-514c-11e8-8d2f-4f64cfab6e66.png">

Which you need to kill the server to stop.

In making the example repo, I came up with the solution in my [`next.config.js`](https://github.com/orta/nextjs-typescript-hrm-bug-example/blob/master/next.config.js):

```js
    if (dev && !isServer) {
      const HMR = config.module.rules.find(r => r.loader === "hot-self-accept-loader")
      if (HMR) {
        HMR.test = /\.(ts|tsx|js|jsx)$/
        HMR.options.extensions = /\.(ts|tsx|js|jsx)$/
      }
    }
```

Which modifies the HMR to support TS.

## Expected Behavior

This should probably be documented somewhere 👍 

Perhaps the with-typescript example could be updated to babel's TypeScript instead, and this can be added to the config file?

## Steps to Reproduce (for bugs)

```
git clone https://github.com/orta/nextjs-typescript-hrm-bug-example
cd nextjs-typescript-hrm-bug-example
yarn
yarn dev
```

Then change `index.tsx`

## Context

n/a

**Comment by herrstucki** [comment]
I haven't checked but I can fix it here too: https://github.com/zeit/next-plugins/pull/162

**Comment by orta** [comment]
That won’t fix it I’m afraid, it should still have the same infinite loop behavior (the example repo is using babel7)

**Comment by herrstucki** [comment]
@orta I didn't mean that it's already fixed there but that I could incorporate your fix into that PR 😃 (which would be the right place to fix it I think).

**Comment by orta** [comment]
Ah, yeah, good thinking 👍

**Comment by herrstucki** [comment]
@orta I can't seem to reproduce your issue though neither in Safari, Chrome, FF – so I'm not sure if a fix is needed 🤔.

Your repro has one (unrelated) problem however: you shouldn't wrap the page in `<html>` and `<body>`. This should be enough:

```jsx
import Head from "next/head"

export default () => (
  <div style={{ textAlign: "center", paddingTop: "90px", fontFamily: "Avenir Next" }}>
    <Head>
      <title>Hi!</title>
    </Head>
    <p>Hello there, please change this text</p>
  </div>
)
```

**Comment by orta** [comment]
Ah dang, I [committed the fix](https://github.com/orta/nextjs-typescript-hrm-bug-example/blob/master/next.config.js#L31-L40) though - wasn't meant to

**Comment by herrstucki** [comment]
Yeah, I noticed 😃 but I don't get the infinite reload loop with or without your fix.

**Comment by DullReferenceException** [comment]
I'm intermittently getting an infinite reload loop with the integration tests. I thought it was due to the changes in my PR at first, but checking out master, I can still reproduce. It makes me wonder if this reload loop even relates to TypeScript or if there's an issue with HMR overall.

**Comment by timneutkens** [comment]
Something similar to the fix @orta described is now in the core and out on next@canary, @orta could you verify if you still run into it?

**Comment by orta** [comment]
I've just given it a shot and I'm afraid I still get infinite loops, video: https://ortastuff.s3.amazonaws.com/vids/peril-next-canary.mov

**Comment by ex3ndr** [comment]
Same here

**Comment by landvibe** [comment]
same here.
next v6.1.1, typescript, only safari

**Comment by timneutkens** [comment]
Will probably be fixed by #4639
